### DISCUSSION TITLE
compiling on windows

### DISCUSSION THREAD

**Discussion by schnapsy** [nohal/nsk_pi#13] [Q&A]
after these instructions :
_git clone https://github.com/nohal/nsk_pi.git
cd nsk_pi.git_
there is no directory "nsk_pi.git"

So i open cd nsk_pi in administrtor mode and paste the instruction :
_ci\appveyor.bat_

and i have these error
_CMake Error at CMakeLists.txt:45 (add_subdirectory):
  add_subdirectory given source "opencpn-libs/api-18" which is not an
  existing directory._

_CMake Error at CMakeLists.txt:59 (string):
  string sub-command REGEX, mode REPLACE needs at least 6 arguments total to
  command_._

**Comment by nohal** [comment]
Run `git submodule update --init`

**Comment by schnapsy** [comment]
(...)           ]
    Génération de code en cours...

_what follow is in RED_ 

C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: La commande "setlocal [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: cmake --build C:/Users/Alain/OCNplugins/nsk_pi/build --parallel 4 --config RelWithDebInfo --target NSK_pi
 [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: if %errorlevel% neq 0 goto :cmEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball
-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmErrorLevel [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: exit /b %1 [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :cmDone [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: if %errorlevel% neq 0 goto :VCEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159,5): e
rror MSB3073: :VCEnd" s'est arrêtée avec le code 1. [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]

_END RED_

Traceback (most recent call last):
  File "C:\Users\Alain\OCNplugins\nsk_pi\build\ldd.py", line 2, in <module>
    lib = glob.glob("app/*/plugins/*.dll")[0]
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
IndexError: list index out of range
Uploading artifact
Warning: CLOUDSMITH_API_KEY is not available, giving up.
Metadata: NSK-0.2-msvc-wx32-10.xml
Tarball: NSK-0.2.3+2305171005.abf231b_msvc-wx32-10-win32.tar.gz
Version: 0.2.3+2305171005.abf231b
Pushing updates to catalog
No $GIT_REPO found, cannot push to git repo.

**Comment by nohal** [comment]
You have built the plugin and have a tarball, so everything works, as far as I can tell. You of course can't push that tarball to my Cloudsmith repository, that is normal.

**Comment by nohal** [comment]
I'm converting this issue to a discussion as it is technically neither a bug report nor a feature request.

**Reply by schnapsy** [reply]
I don't understand your answer.
My question was about the error messages, not about the tarball push.

I have indicated that these error messages are in red.

C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: The command "setlocal [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: cmake --build C:/Users/Alain/OCNplugins/nsk_pi/build --parallel 4 --config RelWithDebInfo --target NSK_pi
[C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
(etc.)

(see previous post)

**Comment by schnapsy** [comment]
I deleted everything and started over.

In order :
_git clone https://github.com/nohal/nsk_pi.git
cd nsk_pi.git
git submodule update --init
ci\appveyor.bat_

After running "ci\appveyor.bat", I get the following error message:
  _Code generation in progress...
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: The command "setlocal [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: cmake --build C:/Users/Alain/OCNplugins/nsk_pi/build --parallel 4 --config RelWithDebInfo --target NSK_pi
  [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: if %errorlevel% neq 0 goto:cmEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073::cmEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: endlocal & call:cmErrorLevel %errorlevel% & goto:cmDone [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball
-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073::cmErrorLevel [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: exit /b %1 [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073::cmDone [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073: if %errorlevel% neq 0 goto:VCEnd [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets(159.5):e
rror MSB3073::VCEnd" exited with code 1. [C:\Users\Alain\OCNplugins\nsk_pi\build\tarball-build.vcxproj]_

**Comment by nohal** [comment]
Here is what I just did:

- Start the VS Command prompt as Administrator

```
**********************************************************************
** Visual Studio 2022 Developer Command Prompt v17.6.0
** Copyright (c) 2022 Microsoft Corporation
**********************************************************************
[vcvarsall.bat] Environment initialized for: 'x86'

C:\Windows\System32>cd \Users\nohal

C:\Users\nohal>git clone --recurse-submodules https://github.com/nohal/nsk_pi
Cloning into 'nsk_pi'...
remote: Enumerating objects: 760, done.
remote: Counting objects: 100% (169/169), done.
remote: Compressing objects: 100% (137/137), done.
remote: Total 760 (delta 62), reused 71 (delta 30), pack-reused 591
Receiving objects: 100% (760/760), 7.09 MiB | 3.84 MiB/s, done.
Resolving deltas: 100% (271/271), done.
Submodule 'opencpn-libs' (https://github.com/leamas/opencpn-libs.git) registered for path 'opencpn-libs'
Cloning into 'C:/Users/nohal/nsk_pi/opencpn-libs'...
remote: Enumerating objects: 1660, done.
remote: Counting objects: 100% (127/127), done.
remote: Compressing objects: 100% (71/71), done.
remote: Total 1660 (delta 70), reused 94 (delta 53), pack-reused 1533
Receiving objects: 100% (1660/1660), 23.68 MiB | 1.62 MiB/s, done.
Resolving deltas: 100% (572/572), done.
Submodule path 'opencpn-libs': checked out '99a92e668254dbd35d5f99f067c4bff95877e14d'

C:\Users\nohal>cd nsk_pi

C:\Users\nohal\nsk_pi>ci\appveyor.bat
Python 3.11.3
Defaulting to user installation because normal site-packages is not writeable
Looking in links: c:\Users\nohal\AppData\Local\Temp\tmp7y7ofbz0
Requirement already satisfied: setuptools in c:\program files\windowsapps\pythonsoftwarefoundation.python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0\lib\site-packages (65.5.0)
Requirement already satisfied: pip in c:\users\nohal\appdata\local\packages\pythonsoftwarefoundation.python.3.11_qbz5n2kfra8p0\localcache\local-packages\python311\site-packages (23.1.2)
Requirement already satisfied: pip in c:\users\nohal\appdata\local\packages\pythonsoftwarefoundation.python.3.11_qbz5n2kfra8p0\localcache\local-packages\python311\site-packages (23.1.2)
--2023-05-17 09:04:16--  https://github.com/wxWidgets/wxWidgets/releases/download/v3.2.2.1/wxWidgets-3.2.2.1-headers.7z
Resolving github.com (github.com)... 20.201.28.151
Connecting to github.com (github.com)|20.201.28.151|:443... connected.
HTTP request sent, awaiting response... 302 Found

....
Building...
....

  -- Installing: C:/Users/nohal/nsk_pi/build/app/files/lib/marnav.lib
  -- Installing: C:/Users/nohal/nsk_pi/build/app/files/lib/cmake/marnav/marnav-targets.cmake
  -- Installing: C:/Users/nohal/nsk_pi/build/app/files/lib/cmake/marnav/marnav-targets-relwithdebinfo.cmake
  -- Installing: C:/Users/nohal/nsk_pi/build/app/files/lib/cmake/marnav/marnav-config.cmake
  -- Installing: C:/Users/nohal/nsk_pi/build/app/files/lib/cmake/marnav/marnav-config-version.cmake
  Building Custom Rule C:/Users/nohal/nsk_pi/CMakeLists.txt
  -- Creating tarball NSK-0.2.3+2305171604.e6b902a_msvc-wx32-10-win32.tar.gz
  -- Computing checksum in NSK-0.2-msvc-wx32-10.xml
  Building Custom Rule C:/Users/nohal/nsk_pi/CMakeLists.txt
'dumpbin' is not recognized as an internal or external command,
operable program or batch file.
Uploading artifact
Warning: CLOUDSMITH_API_KEY is not available, giving up.
Metadata: NSK-0.2-msvc-wx32-10.xml
Tarball: NSK-0.2.3+2305171604.e6b902a_msvc-wx32-10-win32.tar.gz
Version: 0.2.3+2305171604.e6b902a
Pushing updates to catalog
No $GIT_REPO found, cannot push to git repo.

C:\Users\nohal\nsk_pi>
```

That is how it should look on a sane environment

**Reply by schnapsy** [reply]
After trying on 3 different PCs, I finally tried to compile the original plugin and not the one I had modified.
**_And it worked, thanks_**.
Now I need to understand exactly the detail of SignalK sentences I want to add, especially XDR sentences.

**Comment by nohal** [comment]
If I can give you an advice, start with anything else but `XDR`, that sentence and the way the industry implemented it is actually a pure essence of what is wrong with NMEA 0183...
